core/src/main/java/org/apache/iceberg/ManifestFilterManager.java

```diff
@@ -0,0 +1,426 @@
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.CharSequenceSet;
+import org.apache.iceberg.util.CharSequenceWrapper;
+import org.apache.iceberg.util.ManifestFileUtil;
+import org.apache.iceberg.util.StructLikeWrapper;
+import org.apache.iceberg.util.Tasks;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+abstract class ManifestFilterManager<F extends ContentFile<F>> {
+  private static final Logger LOG = LoggerFactory.getLogger(ManifestFilterManager.class);
+  private static final Joiner COMMA = Joiner.on(",");
+
+  protected static class DeleteException extends ValidationException {
+    private final String partition;
+
+    private DeleteException(String partition) {
+      super("Operation would delete existing data");
+      this.partition = partition;
+    }
+
+    public String partition() {
+      return partition;
+    }
+  }
+
+  private final Set<CharSequence> deletePaths = CharSequenceSet.empty();
+  private final Set<StructLikeWrapper> deleteFilePartitions = Sets.newHashSet();
+  private final Set<StructLikeWrapper> dropPartitions = Sets.newHashSet();
+  private Expression deleteExpression = Expressions.alwaysFalse();
+  private boolean hasPathOnlyDeletes = false;
+  private boolean failAnyDelete = false;
+  private boolean failMissingDeletePaths = false;
+  private int duplicateDeleteCount = 0;
+
+  // cache filtered manifests to avoid extra work when commits fail.
+  private final Map<ManifestFile, ManifestFile> filteredManifests = Maps.newConcurrentMap();
+
+  // tracking where files were deleted to validate retries quickly
+  private final Map<ManifestFile, Iterable<F>> filteredManifestToDeletedFiles =
+      Maps.newConcurrentMap();
+
+  protected abstract PartitionSpec spec(int specId);
+  protected abstract void deleteFile(String location);
+  protected abstract ManifestWriter<F> newManifestWriter(PartitionSpec spec);
+  protected abstract ManifestReader<F> newManifestReader(ManifestFile manifest);
+
+  protected void failAnyDelete() {
+    this.failAnyDelete = true;
+  }
+
+  protected void failMissingDeletePaths() {
+    this.failMissingDeletePaths = true;
+  }
+
+  /**
+   * Add a filter to match files to delete. A file will be deleted if all of the rows it contains
+   * match this or any other filter passed to this method.
+   *
+   * @param expr an expression to match rows.
+   */
+  protected void deleteByRowFilter(Expression expr) {
+    Preconditions.checkNotNull(expr, "Cannot delete files using filter: null");
+    invalidateFilteredCache();
```

Comment:
We used to have a flag `filterUpdated` and take it into in `apply`. It seems okay to invalidate eagerly too.